wp-includes/rest-api/endpoints/class-wp-rest-customize-changesets-controller.php

```diff
@@ -94,4 +133,835 @@ public function register_routes() {
+
+		$prepared_post->post_type = $this->post_type;
+
+		$post_id = wp_insert_post( wp_slash( (array) $prepared_post ), true );
+
+		if ( is_wp_error( $post_id ) ) {
+
+			if ( 'db_insert_error' === $post_id->get_error_code() ) {
+				$post_id->add_data( array(
+					'status' => 500,
+				) );
+			} else {
+				$post_id->add_data( array(
+					'status' => 400,
+				) );
+			}
+
+			return $post_id;
+		}
+
+		$post = get_post( $post_id );
+
+		/**
+		 * Fires after a changeset post is created or updated via the REST API.
+		 *
+		 * @since 4.?.?
+		 *
+		 * @param WP_Post         $post     Inserted or updated post object.
+		 * @param WP_REST_Request $request  Request object.
+		 * @param bool            $creating True when creating a post, false when updating.
+		 */
+		do_action( "rest_insert_{$this->post_type}", $post, $request, true );
+
+		$fields_update = $this->update_additional_fields_for_object( $post, $request );
+
+		if ( is_wp_error( $fields_update ) ) {
+			return $fields_update;
+		}
+
+		$request->set_param( 'context', 'edit' );
+
+		$response = $this->prepare_item_for_response( $post, $request );
+		$response = rest_ensure_response( $response );
+
+		$response->set_status( 201 );
+		$response->header( 'Location', rest_url( sprintf( '%s/%s/%d', $this->namespace, $this->rest_base, $post->post_name ) ) );
+
+		return $response;
+	}
+
+	/**
+	 * Prepares a customize changeset for create or update.
+	 *
+	 * @since 4.?.?
+	 * @access protected
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return stdClass|WP_Error Post object or WP_Error.
+	 */
+	protected function prepare_item_for_database( $request ) {
+		$prepared_post = new stdClass;
+
+		$manager = $this->ensure_customize_manager( $request['uuid'] );
+		$prepared_post->ID = $manager->changeset_post_id();
+		$prepared_post->post_name = $request['uuid'];
+
+		// Post title.
+		if ( isset( $request['title'] ) ) {
+			if ( is_string( $request['title'] ) ) {
+				$prepared_post->post_title = $request['title'];
+			} elseif ( ! empty( $request['title']['raw'] ) ) {
+				$prepared_post->post_title = $request['title']['raw'];
+			}
+		}
+
+		// Settings.
+		if ( isset( $request['settings'] ) ) {
+			$settings = array();
+
+			if ( ! is_array( $request['settings'] ) ) {
+				return new WP_Error( 'invalid_customize_changeset_data', __( 'Invalid customize changeset data.' ), array(
+					'status' => 400,
+				) );
+			}
+			foreach ( $request['settings'] as $setting_id => $params ) {
+
+				$setting = $manager->get_setting( $setting_id );
+				if ( ! $setting ) {
+					return new WP_Error( 'invalid_customize_changeset_data', __( 'Invalid setting.' ), array(
+						'status' => 400,
+					) );
+				}
+				if ( ! $setting->check_capabilities() ) {
+					return new WP_Error( 'rest_forbidden', __( 'Sorry, you are not allowed to edit some of the settings.' ), array(
+						'status' => 403,
+					) );
+				}
+				$settings[ $setting_id ] = array(
+					'value' => $params['value'],
```

Comment:
What about allowing the user to attach other params to settings? Currently this is somewhat ad hoc, to be honest in how the core is implemented, but the intention in the core implementation was that additional metadata could be attached to settings in a changeset, although some properties should be reserved like `user_id`, `type`, and `starter_content`.